src/librustc_trans/adt.rs

```diff
@@ -1262,45 +761,52 @@ fn compute_struct_field_offsets<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
 /// a two-element struct will locate it at offset 4, and accesses to it
 /// will read the wrong memory.
 fn build_const_struct<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
-                                st: &Struct<'tcx>, vals: &[ValueRef])
+                                offset_after_field: &[layout::Size],
+                                vals: &[ValueRef],
+                                packed: bool)
                                 -> Vec<ValueRef> {
-    assert_eq!(vals.len(), st.fields.len());
+    assert_eq!(vals.len(), offset_after_field.len());
 
-    let target_offsets = compute_struct_field_offsets(ccx, st);
+    if vals.len() == 0 {
+        return Vec::new();
+    }
 
     // offset of current value
     let mut offset = 0;
     let mut cfields = Vec::new();
-    for (&val, target_offset) in vals.iter().zip(target_offsets) {
-        if !st.packed {
+    for (&val, target_offset) in
+        vals.iter().zip(
+        offset_after_field.iter().map(|i| i.bytes())
+    ) {
```

Comment:
You can put the second iterator in a `target_offsets` variable to recover the original look.